man/fatsort.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.14.
.TH FATSORT "1" "August 2020" "fatsort 1.6.3" "User Commands"
.SH NAME
fatsort \- FAT sorting tool
.SH SYNOPSIS
.B fatsort
[\fI\,OPTIONS\/\fR] \fI\,DEVICE\/\fR
.SH DESCRIPTION
FATSort sorts directory structures of FAT file systems (FAT12, FAT16, FAT32 and exFAT). Many MP3 hardware players don't sort files automatically but play them in the  order they were transferred to the device. FATSort can help here.
.PP
Sort options:
.TP
\fB\-a\fR
Use ASCIIbetical order for sorting
.TP
\fB\-c\fR
Ignore case of file names (default for most locales)
.TP
\fB\-I\fR PFX
Ignore file name PFX
.TP
\fB\-n\fR
Natural order sorting
.HP
\fB\-o\fR FLAG Sort order of files where FLAG is one of
.IP
d : directories first (default)
.IP
f : files first
.IP
a : files and directories are not differentiated
.TP
\fB\-r\fR
Sort in reverse order
.TP
\fB\-R\fR
Sort in random order
.IP
This options makes all other sort options obsolete.
.TP
\fB\-t\fR
Sort by last modification date and time
.PP
The following options can be specified multiple times
to select which directories shall be sorted:
.TP
\fB\-d\fR DIR
Sort directory DIR only
.TP
\fB\-D\fR DIR
Sort directory DIR and all subdirectories
.TP
\fB\-x\fR DIR
Don't sort directory DIR
.TP
\fB\-X\fR DIR
Don't sort directory DIR and its subdirectories
.PP
The following options can be specified multiple times
to select which directories shall be sorted using
POSIX.2 extended regular expressions:
.TP
\fB\-e\fR RE
Only sort directories that match regular expression RE
.TP
\fB\-E\fR RE
Don't sort directories that match regular expression RE
.PP
However, options \fB\-e\fR and \fB\-E\fR may not be used simultaneously with
options \fB\-d\fR, \fB\-D\fR, \fB\-x\fR and \fB\-X\fR.
.TP
\fB\-L\fR LOC
Use the locale LOC instead of the locale from the environment variables
.PP
More options:
.TP
\fB\-l\fR
Print current order of files only
.TP
\fB\-i\fR
Print file system information only
.TP
\fB\-f\fR
Force sorting even if file system is mounted
.HP
\fB\-h\fR, \fB\-\-help\fR
.IP
Print some help
.HP
\fB\-v\fR, \fB\-\-version\fR
.IP
Print version information
.TP
\fB\-q\fR
Be quiet
.PP
DEVICE must be a FAT12, FAT16, FAT32 or exFAT file system.
.PP
WARNING: THE FILESYSTEM MUST BE CONSISTENT (NO FILESYSTEM ERRORS).
PLEASE BACKUP YOUR DATA BEFORE USING FATSORT. RISK OF CORRUPT FILESYSTEM!
FATSORT USER ASSUMES ALL RISK. FATSORT WILL NOT BE HELD LIABLE FOR DATA LOSS!
.SH EXAMPLES
.TP
fatsort /dev/sda
Sort /dev/sda.
.TP
fatsort \-n /dev/sdb1
Sort /dev/sdb1 with natural order.
.SH AUTHOR
Written by Boris Leidner.
.SH "REPORTING BUGS"
Report bugs to <fatsort@formenos.de>.
.SH COPYRIGHT
Copyright \(co 2004\-2019 Boris Leidner.
License GPLv2: GNU GPL version 2 (see LICENSE.txt)
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


CHANGES.md
# Changelog

## v1.6.3 (August  8th, 2020)
* fatsort release file now includes precompiled makefile
* fixed support for macOS (thanks to Max for the fix)
* declared OPT_LOCALE as extern
* fixed uninitialized variable
* avoided some string truncation compiler warnings
* implemented platform independent transliteration for invalid multi-byte characters during conversion
* test makefile now returns error code

## v1.6.2 (November 29, 2019)
- FIX: multiple endianness issues with exFAT
- now using /proc/self/mounts instead of /etc/mtab to check whether filesystem is mounted

## v1.6.1 (November 17, 2019)
- FIX: missing check if directory entries are of type file
- option -l now considers options -d, -D, -x, -X, -e and -E
- added check for dirty volumes
- added new test cases

## v1.6.0 (November 16, 2019)
- exFAT is now supported!
- directory entries are now only written back if reordering happened
- FIX: memory leak with regular expression lists
- code rewrite to make additional platforms possible in future

## v1.5.0 (September 19, 2018)
- added options -e and -E to select directories with regular expressions
- FIX: possible memory leak for `OPT_LOCALE`
- FIX: bug affecting case sensitive sorting with locales

## v1.4.2 (August 06, 2017)
- FIX: sorting with option -m on big endian architectures is now working as intended
- reproducible docker tests added for i386, amd64, ppc64le
- added clean target for manpage
- updated tests

## v1.4.1 (July 29, 2017)
- fixed memory leak when providing multiple locales
- fixed test tc\_check\_consistent\_version
- added //TRANSLIT to `iconv_open` so iconv will no longer fail when a character cannot be converted
- added missing test cases

## v1.4 (July 28, 2017)
- added option -L to specify an alternate locale to be used
- added warning if C locale is used
- added valgrind tests
- fixed possible memory leak found with valgrind

## v1.3.1 (July 19, 2017)
- increased maximum cluster size to 64kb
- fixed descriptions for options -I, -o, -d, -D, -x and -X in manpage
- replaced some `fclose` with `fs_close`
- changed iconv error to warning
- fixed a test case

## v1.3 (February 28, 2014)
- added option -a for ASCIIbetical sorting
- added test cases for option -a

## v1.2 (July 12, 2013)
- fatsort now sychronizes filesystems after sorting each directory
- added long options --version and --help
- updated version and help output for help2man
- added Makefile for manpage
- minor changes

## v1.1.1 (June 09, 2013)
- fixed character conversion for non-UTF-8 systems

## v1.1 (May 23, 2013)
- New option -t to sort by last modification date time. Thanks to Charles Anthony for the patch.
- More and updated tests
- Minor changes

## v1.0.1 (May 17, 2013)
- FIX: big endian issue for FAT12/16 in function getFATEntry
- FIX: big endian issue for `BS_SecPerClus`
- FIX: Stack overflow in cmpentries
- FIX: checkLongDirEntries now ignores deleted entries
- added check for maximum clusters according to MS Q184006
- considered -Wextra warnings
- make fatsort detect cluster chain loops
- implemented regression tests
- updated fatsort help information
- Option combination -l and -m now also print long AND short file names
- changed `DE_FREE` to `u_char`
- additional cluster size check
- Optimized calculation of file system information
- Minor changes and code cleaning

## v1.0 (March 11, 2013)
- FIX: exit critical sections on error
- FIX: Natural order sorting works again (without collation ordering though)
- ENHANCEMENT: Added support for FAT12
- Added checks for broken clusterchains
- Added cluster usage information for option -i
- Checksums for LongDirEntries are now verified
- Added integrity check that verifies whether all FATs are the same

## v0.9.17 (April 4, 2012)
- Collation ordering for the current locale is now considered during sorting
- Fixed some random bugs found with a static code analysis tools

## v0.9.16 (December July, 2011)
- Updated address in GPL license

## v0.9.15 (August 15, 2011)
- Multiple directories can now be included and excluded for sorting using options -d, -D, -x, and -X. Thanks to Renaud Métrich for the idea.
- Added option -I which allows to ignore filename prefixes during sorting. Thanks to Matthew Altus for the idea.
- Added support for memory leak detection in debug modes 2-4
- Fixed a bug in function randomizeDirEntryList. Thanks to Renaud Métrich for reporting the bug.
- Updated Makefiles for better Max OS X support. Thanks to Tom Freiberger for the Makefiles.
- Minor change to remove a compiler warning in FAT\_fs.c.
- Minor change to remove a redundant option check.

## v0.9.14.1 (June 11, 2010)
- Check for last directory entry in FAT\_fs.c was fixed. thanks to Peter De Wachter for the hint.

## v0.9.14 (May 14, 2010)
- Included a patch for an iconv issue under Mac and FreeBSD. Thanks to Max Campos.

## v0.9.13.1 (December 12, 2009)
- Fixed a pointer initialization in insertLongDirEntryList

## v0.9.13 (December 23, 2009)
- Added random order sorting (option -R)
- Updated manpage
- manpage is now installed via make install

## v0.9.12 (September 29, 2009)
- locale is now set to locale of environment

## v0.9.11 (September 27, 2009)
- segmentation fault in parseClusterChain fixed
- LFNs are now correctly converted from UTF-16 to UTF-8

## v0.9.10 (May 12, 2008)
- Options -d and -D added
- natural order comparison algorithm reimplemented

## v0.9.9.1 (February 12, 2008)
- Small Makefile update

## v0.9.9 (November 28, 2008)
- Martin Pool's natural order comparison algorithm is now included into FATSort
- Makefile patches

## v0.9.8.3 (July 08, 2008)
- bug in `check_mounted` fixed

## v0.9.8.2 (June 11, 2008)
- minor fixes in header files and make files
- modified README

## v0.9.8.1 (June 10, 2008)
- some minor fixes. Thanks to Till Maas.

## v0.9.8 (Septembe 25, 2007)
- splitted fatsort into several source files
- implemented exclusive file lock (not mandatory)
- signals are blocked while writing to file (implemented critial sections)

src/fatsort.c
/*
	FATSort, utility for sorting FAT directory structures
	Copyright (C) 2018 Boris Leidner <fatsort(at)formenos.de>

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

/*
	This file contains the main function of fatsort.
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdarg.h>
#include <assert.h>
#include <errno.h>
#include <locale.h>
#include <time.h>
#include <inttypes.h>

// project includes
#include "endianness.h"
#include "sig.h"
#include "FAT_fs.h"
#include "options.h"
#include "errors.h"
#include "sort.h"
#include "clusterchain.h"
#include "misc.h"
#include "mallocv.h"

// program information
#define INFO_PROGRAM		"fatsort"
#define INFO_VERSION		"1.6.3"
#define INFO_AUTHOR		"Written by Boris Leidner.\n"
#define INFO_COPYRIGHT		"Copyright (C) 2004-2019 Boris Leidner.\n"
#define INFO_LICENSE		"License GPLv2: GNU GPL version 2 (see LICENSE.txt)\n" \
				"This is free software: you are free to change and redistribute it.\n" \
				"There is NO WARRANTY, to the extent permitted by law.\n"
#define INFO_DESCRIPTION	"FATSort sorts directory structures of FAT file systems (FAT12, FAT16, FAT32 and exFAT). " \
				"Many MP3 hardware players don't sort files automatically " \
				"but play them in the  order they were transferred to the " \
				"device. FATSort can help here.\n"

#define INFO_USAGE		"Usage: fatsort [OPTIONS] DEVICE\n" \
				"\n" \
				"Sort options:\n\n" \
				"\t-a\tUse ASCIIbetical order for sorting\n\n" \
				"\t-c\tIgnore case of file names (default for most locales)\n\n" \
				"\t-I PFX\tIgnore file name PFX\n\n" \
				"\t-n\tNatural order sorting\n\n" \
				"\t-o FLAG\tSort order of files where FLAG is one of\n\n" \
				"\t\t\td : directories first (default)\n\n" \
				"\t\t\tf : files first\n\n" \
				"\t\t\ta : files and directories are not differentiated\n\n" \
				"\t-r\tSort in reverse order\n\n" \
				"\t-R\tSort in random order\n\n" \
				"\t\tThis options makes all other sort options obsolete.\n\n" \
				"\t-t\tSort by last modification date and time\n\n" \
				"The following options can be specified multiple times\n" \
				"to select which directories shall be sorted:\n\n" \
				"\t-d DIR\tSort directory DIR only\n\n" \
				"\t-D DIR\tSort directory DIR and all subdirectories\n\n" \
				"\t-x DIR\tDon't sort directory DIR\n\n" \
				"\t-X DIR\tDon't sort directory DIR and its subdirectories\n\n" \
				"The following options can be specified multiple times\n" \
				"to select which directories shall be sorted using\n" \
				"POSIX.2 extended regular expressions:\n\n" \
				"\t-e RE\tOnly sort directories that match regular expression RE\n\n" \
				"\t-E RE\tDon't sort directories that match regular expression RE\n\n" \
				"However, options -e and -E may not be used simultaneously with\n" \
				"options -d, -D, -x and -X.\n\n" \
				"\t-L LOC\tUse the locale LOC instead of the locale from the environment variables\n\n" \
				"More options:\n\n" \
				"\t-l\tPrint current order of files only\n\n" \
				"\t-i\tPrint file system information only\n\n" \
				"\t-f\tForce sorting even if file system is mounted\n\n" \
				"\t-h, --help\n\n" \
				"\t\tPrint some help\n\n" \
				"\t-v, --version\n\n" \
				"\t\tPrint version information\n\n" \
				"\t-q\tBe quiet\n\n" \
				"DEVICE must be a FAT12, FAT16, FAT32 or exFAT file system.\n\n" \
				"WARNING: THE FILESYSTEM MUST BE CONSISTENT (NO FILESYSTEM ERRORS).\n" \
				"PLEASE BACKUP YOUR DATA BEFORE USING FATSORT. RISK OF CORRUPT FILESYSTEM!\n" \
				"FATSORT USER ASSUMES ALL RISK. FATSORT WILL NOT BE HELD LIABLE FOR DATA LOSS!\n" \
								"\n" \
				"Examples:\n" \
				"\tfatsort /dev/sda\t\tSort /dev/sda.\n" \
				"\tfatsort -n /dev/sdb1\t\tSort /dev/sdb1 with natural order.\n" \
				"\n" \
				"Report bugs to <fatsort@formenos.de>.\n"

#define INFO_OPTION_VERSION	INFO_PROGRAM " " INFO_VERSION "\n" \
				"\n" \
				INFO_COPYRIGHT \
				INFO_LICENSE \
				"\n" \
				INFO_AUTHOR

#define INFO_OPTION_HELP	INFO_DESCRIPTION \
				"\n" \
				INFO_USAGE


int32_t printFSInfo(char *filename) {
/*
	print file system information
*/

	assert(filename != NULL);

	uint32_t value, clen;
	uint32_t usedClusters, badClusters;
	uint32_t i;
	struct sClusterChain *chain;

	struct sFileSystem fs;

	printf("\t- File system information -\n");

	if (openFileSystem(filename, FS_MODE_RO, &fs)) {
		myerror("Failed to open file system!");
		return -1;
	}

	if (fs.FATType != 64) { // VFAT
		usedClusters=0;
		badClusters=0;
		for (i=2; i<fs.clusters+2; i++) {
			getFATEntry(&fs, i, &value);
			if ((value & 0x0FFFFFFF) != 0) usedClusters++;
			if (fs.FATType == FATTYPE_FAT32) {
				if ((value & 0x0FFFFFFF) == 0x0FFFFFF7) badClusters++;
			} else if (fs.FATType == FATTYPE_FAT16) {
				if (value == 0x0000FFF7) badClusters++;
			} else if (fs.FATType == FATTYPE_FAT12) {
				if (value == 0x00000FF7) badClusters++;
			} else {
				myerror("Unsupported FAT type!");
				return -1;
			}
		}
	} else { // exFAT
		usedClusters=fs.allocatedClusters;
		badClusters=0;
		for (i=2; i<fs.clusters+2; i++) {
			getFATEntry(&fs, i, &value);
			if (value == 0xFFFFFFF7) badClusters++;
		}
	}

	// feature: print volume label
	printf("Device:\t\t\t\t\t%s\n", fs.path);
	printf("Type:\t\t\t\t\t");
	if (fs.FATType == 12) {
		printf("FAT12");
	} else if (fs.FATType == 16) {
		printf("FAT16");
	} else if (fs.FATType == 32) {
		printf("FAT32");
	} else if (fs.FATType == 64) {
		printf("exFAT %u.%u", fs.bs.xxFATxx.exFAT.version.major, fs.bs.xxFATxx.exFAT.version.minor);
	}
	printf("\n");
	printf("Sector size:\t\t\t\t%d bytes\n", (int) fs.sectorSize);
	printf("FAT size:\t\t\t\t%d sectors (%d bytes)\n", (int) fs.FATSize, (int) (fs.FATSize * fs.sectorSize));
	printf("Number of FATs:\t\t\t\t%d %s\n", fs.FATCount, (checkFATs(&fs) ? "- WARNING: FATs are different!" : ""));
	printf("Cluster size:\t\t\t\t%d bytes\n", (int) fs.clusterSize);
	printf("Max. cluster chain length:\t\t%u clusters\n", (int) fs.maxClusterChainLength);
	printf("Data clusters (total / used / bad):\t%d / %d / %d\n", (int) fs.clusters, (int) usedClusters, (int) badClusters);
	printf("FS size:\t\t\t\t%.2f MiBytes\n", (float) fs.FSSize / (1024.0*1024));

	if (fs.FATType == FATTYPE_FAT32) {
		if (getFATEntry(&fs, SwapInt32(fs.bs.xxFATxx.FAT12_16_32.FATxx.FAT32.BS_RootClus), &value) == -1) {
			myerror("Failed to get FAT entry!");
			closeFileSystem(&fs);
			return -1;
		}
		printf("FAT32 root first cluster:\t\t0x%x\nFirst cluster data offset:\t\t0x%lx\nFirst cluster FAT entry:\t\t0x%x\n",
			(unsigned int) SwapInt32(fs.bs.xxFATxx.FAT12_16_32.FATxx.FAT32.BS_RootClus),
			(unsigned long) getClusterOffset(&fs, SwapInt32(fs.bs.xxFATxx.FAT12_16_32.FATxx.FAT32.BS_RootClus)), (unsigned int) value);
	} else if (fs.FATType == FATTYPE_FAT12) {
		printf("FAT12 root directory Entries:\t\t%u\n", SwapInt16(fs.bs.xxFATxx.FAT12_16_32.BS_RootEntCnt));
	} else if (fs.FATType == FATTYPE_FAT16) {
		printf("FAT16 root directory Entries:\t\t%u\n", SwapInt16(fs.bs.xxFATxx.FAT12_16_32.BS_RootEntCnt));
	} else if (fs.FATType == FATTYPE_EXFAT) {
		printf("Root directory cluster:\t\t\t%u (at 0x%"PRIx64")\n",
				SwapInt32(fs.bs.xxFATxx.exFAT.rootdir_cluster),
				(uint64_t) getClusterOffset(&fs, SwapInt32(fs.bs.xxFATxx.exFAT.rootdir_cluster)));
		printf("Allocation Bitmap cluster:\t\t%u (at 0x%"PRIx64")\n",
						fs.allocBitmapFirstCluster,
						(uint64_t) getClusterOffset(&fs, fs.allocBitmapFirstCluster));
	}

	if (OPT_MORE_INFO) {
		printf("\n\t- FAT -\n");
		printf("Cluster \tFAT entry\tChain length\n");
		for (i=0; i<fs.clusters+2; i++) {
			getFATEntry(&fs, i, &value);

			clen=0;
			if ((value & 0x0FFFFFFF ) != 0) {
				if ((chain=newClusterChain()) == NULL) {
					myerror("Failed to generate new ClusterChain!");
					return -1;
				}
				clen=getClusterChain(&fs, i, chain);
				freeClusterChain(chain);
			}

			printf("%08x\t%08x\t%u\n", i, value, clen);

		}

	}

	closeFileSystem(&fs);

	return 0;

}

int main(int argc, char *argv[]) {
/*
	parse arguments and options and start sorting
*/

	char *locale;

	// initialize rng
	srand(time(0));

	// initialize blocked signals
	init_signal_handling();
	char *filename;

	if (parse_options(argc, argv) == -1) {
		myerror("Failed to parse options!");
		return -1;
	}

	// use locale from environment or option
	locale=setlocale(LC_ALL, OPT_LOCALE);
	if (locale == NULL) {
		myerror("Could not set locale!");
		return -1;
	} else if (strncmp(locale, "C", 1) == 0){
		myerror("WARNING: The C locale does not support all multibyte characters!");
	}

	if (OPT_HELP) {
		printf(INFO_OPTION_HELP);
		return 0;
	} else if (OPT_VERSION) {
		printf(INFO_OPTION_VERSION);
		return 0;
	} else if (optind < argc -1) {
		myerror("Too many arguments!");
		myerror("Use -h for more help.");
		return -1;
	} else if (optind == argc) {
		myerror("Device must be given!");
		myerror("Use -h for more help.");
		return -1;
	}

	filename=argv[optind];

	if (OPT_INFO) {
		//infomsg(INFO_HEADER "\n\n");
		if (printFSInfo(filename) == -1) {
			myerror("Failed to print file system information");
			return -1;
		}
	} else {
		//infomsg(INFO_HEADER "\n\n");
		if (sortFileSystem(filename) == -1) {
			myerror("Failed to sort file system!");
			return -1;
		}
	}

	freeOptions();

	// report mallocv debugging information
	REPORT_MEMORY_LEAKS

	return 0;
}
